### PR TITLE
Feature - Consensus Message Handler Election

### PR DESCRIPTION
**Pull request by georgiyazovaliiski**

### CODE REVIEW

File: app/persistence/message/message.go [hunk 1]
```diff
@@ -2,15 +2,39 @@ package message
...
+	Fee                  string
+	Signature            string
+	Hash                 string
+	SignerAddress        string
+	TransactionTimestamp string
+}
+
+type ByTimestamp []TransactionMessage
+
+func (tm ByTimestamp) Len() int {
```
Comment (failfmi) [inline app/persistence/message/message.go:22]: Do you need this func and the next two?

File: app/process/handler/consensus-message/consensus-message.go [hunk 2]
```diff
@@ -89,21 +111,95 @@ func (cmh ConsensusMessageHandler) handlePayload(payload []byte) error {
...
 	return nil
 }
 
+func (cmh ConsensusMessageHandler) transactionSent() bool {
+	// TODO: Implement Transaction status check
+	return false
+}
+
+func (cmh ConsensusMessageHandler) findMyPosition(messages []message.TransactionMessage) int {
+	sort.Sort(message.ByTimestamp(messages))
```
Comment (failfmi) [inline app/process/handler/consensus-message/consensus-message.go:194]: Retrieve from db ordered by timestamp instead.

File: app/process/handler/consensus-message/consensus-message.go [hunk 3]
```diff
@@ -89,21 +111,95 @@ func (cmh ConsensusMessageHandler) handlePayload(payload []byte) error {
...
+		Hash:      hexHash,
+		EthTxHash: "0x12345",
+	}
+
+	_, err = proto.Marshal(tasm) // topicMessageBytes
+	if err != nil {
+		log.Error("Could not marshal Topic Aggregated Signatures Message [%s] - [%s]", tasm, err)
+	}
+
+	log.Infof("I should send the message! - [%v]", tasm)
```
Comment (failfmi) [inline app/process/handler/consensus-message/consensus-message.go:179]: This looks more like a debug log.

File: app/process/handler/consensus-message/consensus-message.go [hunk 4]
```diff
@@ -89,21 +111,95 @@ func (cmh ConsensusMessageHandler) handlePayload(payload []byte) error {
...
+	log.Infof("DEADLINE: [%v]", deadline)
+
+	for now.Before(deadline) {
+		if cmh.transactionSent() {
+			return nil
+		}
+		now = time.Now()
+		time.Sleep(10 * time.Second)
+	}
+	log.Infof("I should submit the transaction now!")
```
Comment (failfmi) [inline app/process/handler/consensus-message/consensus-message.go:167]: This looks more like a debug log.

File: app/process/handler/consensus-message/consensus-message.go (hunk 2 above)
Comment (georgiyazovaliiski) [reply app/process/handler/consensus-message/consensus-message.go:194]: Timestamp gets saved as a string, due to the logical flow we chose for the previous steps

File: app/process/handler/consensus-message/consensus-message.go (hunk 3 above)
Comment (georgiyazovaliiski) [reply app/process/handler/consensus-message/consensus-message.go:179]: Yes.

File: app/process/handler/consensus-message/consensus-message.go (hunk 4 above)
Comment (georgiyazovaliiski) [reply app/process/handler/consensus-message/consensus-message.go:167]: Yes.

File: app/process/handler/consensus-message/consensus-message.go [hunk 5]
```diff
@@ -89,21 +111,95 @@ func (cmh ConsensusMessageHandler) handlePayload(payload []byte) error {
...
+	requiredSigCount := len(cmh.validAddresses)/2 + len(cmh.validAddresses)%2
+	log.Infof("Required signatures: [%v]", requiredSigCount)
+
+	txSignatures, err = cmh.repository.GetByTransactionId(m.TransactionId, hexHash)
+	if err != nil {
+		return errors.New(fmt.Sprintf("Could not retrieve Transaction Signatures for Transaction [%s]", m.TransactionId))
+	}
+
+	if len(txSignatures) < requiredSigCount {
+		log.Warnf(fmt.Sprintf("Insignificant amount of Transaction Signatures for Transaction [%s] - [%d] signatures", m.TransactionId, len(txSignatures)))
```
Comment (Daniel-K-Ivanov) [inline app/process/handler/consensus-message/consensus-message.go:143]: If the collected signatures are not enough **yet**, I think that we should log info not warning. It is normal to have a period in which signatures are not the required amount

File: config/application.yml [hunk 6]
```diff
@@ -21,11 +21,17 @@ hedera:
...
       topic_id:
-      polling_interval:
+      polling_interval: 5
     consensus-message:
-      addresses:
\ No newline at end of file
+      topic_id:
+      addresses:
+        -
+      send_deadline: 5
```
Comment (Daniel-K-Ivanov) [inline config/application.yml:37]: imo, this parameter should be in seconds

File: app/process/handler/consensus-message/consensus-message.go [hunk 7]
```diff
@@ -89,21 +111,95 @@ func (cmh ConsensusMessageHandler) handlePayload(payload []byte) error {
...
+}
+
+func (cmh ConsensusMessageHandler) findMyPosition(messages []message.TransactionMessage) int {
+	sort.Sort(message.ByTimestamp(messages))
+	for i := 0; i < len(messages); i++ {
+		if messages[i].SignerAddress == cmh.operatorAddress {
+			return i
+		}
+	}
+	return -1
```
Comment (Daniel-K-Ivanov) [inline app/process/handler/consensus-message/consensus-message.go:200]: if the operator is not amongst signers, I think that it would be more appropriate if we return an error and check for that error instead of `-1`

File: app/process/handler/consensus-message/consensus-message.go [hunk 8]
```diff
@@ -89,21 +111,95 @@ func (cmh ConsensusMessageHandler) handlePayload(payload []byte) error {
...
+		log.Warnf(fmt.Sprintf("Insignificant amount of Transaction Signatures for Transaction [%s] - [%d] signatures", m.TransactionId, len(txSignatures)))
+		return nil
+	}
+
+	pos := cmh.findMyPosition(txSignatures)
+	if pos == -1 {
+		return errors.New(fmt.Sprintf("Operator is not amongst the potential leaders - [%v]", pos))
+	}
+
+	log.Infof("My position: [%v]", pos)
```
Comment (Daniel-K-Ivanov) [inline app/process/handler/consensus-message/consensus-message.go:152]: this log does not have any context for the transaction

File: app/process/handler/consensus-message/consensus-message.go [hunk 9]
```diff
@@ -89,21 +111,95 @@ func (cmh ConsensusMessageHandler) handlePayload(payload []byte) error {
...
+
+	now := time.Now()
+	deadline := now.Add(time.Minute * time.Duration(cmh.deadline*pos))
+
+	log.Infof("NOW: [%v]", now)
+	log.Infof("DEADLINE: [%v]", deadline)
```
Comment (Daniel-K-Ivanov) [inline app/process/handler/consensus-message/consensus-message.go:158]: the logs do not have any transaction context

File: app/process/handler/consensus-message/consensus-message.go (hunk 3 above)
Comment (Daniel-K-Ivanov) [inline app/process/handler/consensus-message/consensus-message.go:179]: What if this operator was the 2nd leader and the 1st one submitted it?
Once someone submits the TransactionHash to the topicID, this flow should be able to figure out that the transaction was already submitted and should stop the leader election for this tx (`leaderElection.stop(txId)`)

**Review by Daniel-K-Ivanov — CHANGES_REQUESTED** [review]

File: app/domain/repositories/message.go [hunk 10]
```diff
@@ -3,6 +3,7 @@ package repositories
 import "github.com/limechain/hedera-eth-bridge-validator/app/persistence/message"
 
 type MessageRepository interface {
-	GetTransaction(txId, signature string) ([]message.TransactionMessage, error)
+	GetTransaction(txId, signature, hash string) ([]message.TransactionMessage, error)
 	Create(message *message.TransactionMessage) error
+	GetByTransactionId(txId string, txHash string) ([]message.TransactionMessage, error)
```
[the remark this answers is not in the document]
Comment (georgiyazovaliiski) [reply app/domain/repositories/message.go:8]: Go does not support method overloading

File: app/process/handler/consensus-message/consensus-message.go [hunk 11]
```diff
@@ -6,32 +6,54 @@ import (
...
+		hederaNodeClient: hederaNodeClient,
+		topicID:          topicID,
 	}
 }
 
 func (cmh ConsensusMessageHandler) Handle(payload []byte) {
+	go cmh.errorHandler(payload)
+}
+
+func (cmh ConsensusMessageHandler) errorHandler(payload []byte) {
```
[the remark this answers is not in the document]
Comment (georgiyazovaliiski) [reply app/process/handler/consensus-message/consensus-message.go:56]: I want to be able to return an error and log it in a single point of output, but keep it in a separate goroutine.

File: app/persistence/message/leader-election.go [hunk 12]
```diff
@@ -0,0 +1,13 @@
+package message
```
Comment (Daniel-K-Ivanov) [inline app/persistence/message/leader-election.go:1]: I think that this code is not related to `persistence` in any way. It is the business logic of the consensus-message handler, thus it should be in that package imo

File: app/process/handler/consensus-message/consensus-message.go [hunk 13]
```diff
@@ -124,84 +124,84 @@ func (cmh ConsensusMessageHandler) handlePayload(payload []byte) error {
...
+	_, err = cmh.hederaNodeClient.SubmitTopicConsensusMessage(cmh.topicID, topicMessageBytes)
+	if err != nil {
+		log.Error("Could not submit Topic Consensus Message [%s] - [%s]", topicMessageBytes, err)
+	}
+}
 
-	return nil
+func (cmh ConsensusMessageHandler) enoughSignaturesCollected(txSignatures []message.TransactionMessage, transactionId string) bool {
+	requiredSigCount := len(cmh.operatorsEthAddresses)/2 + len(cmh.operatorsEthAddresses)%2
+	log.Infof("Required signatures: [%v]", requiredSigCount)
```
Comment (Daniel-K-Ivanov) [inline app/process/handler/consensus-message/consensus-message.go:178]: this log is maybe a leftover for debug purposes only?
If you think that this info is meaningful (I do), you should put it in the log with tx context (below on line 181)

File: app/process/handler/consensus-message/consensus-message.go [hunk 14]
```diff
@@ -6,32 +6,54 @@ import (
...
+	topicID, err := hedera.TopicIDFromString(config.LoadConfig().Hedera.Handler.ConsensusMessage.TopicId)
+	if err != nil {
+		log.Fatal("Invalid topic id: [%v]", config.LoadConfig().Hedera.Handler.ConsensusMessage.TopicId)
+	}
+
 	return &ConsensusMessageHandler{
-		repository:     r,
-		validAddresses: config.LoadConfig().Hedera.Handler.ConsensusMessage.Addresses,
+		repository:            r,
+		operatorsEthAddresses: config.LoadConfig().Hedera.Handler.ConsensusMessage.Addresses,
```
Comment (failfmi) [inline app/process/handler/consensus-message/consensus-message.go:44]: pass config from main instead of loading it

File: app/process/handler/consensus-message/consensus-message.go [hunk 15]
```diff
@@ -6,32 +6,54 @@ import (
...
+		operatorAddress:       eth.PrivateToPublicKeyToAddress(config.LoadConfig().Hedera.Client.Operator.EthPrivateKey).String(),
+		deadline:              config.LoadConfig().Hedera.Handler.ConsensusMessage.SendDeadline,
```
Comment (failfmi) [inline app/process/handler/consensus-message/consensus-message.go:46]: pass config from main instead of loading it

File: app/process/handler/consensus-message/consensus-message.go [hunk 16]
```diff
@@ -6,32 +6,54 @@ import (
...
+	if err != nil {
+		log.Fatal("Invalid topic id: [%v]", config.LoadConfig().Hedera.Handler.ConsensusMessage.TopicId)
+	}
+
 	return &ConsensusMessageHandler{
-		repository:     r,
-		validAddresses: config.LoadConfig().Hedera.Handler.ConsensusMessage.Addresses,
+		repository:            r,
+		operatorsEthAddresses: config.LoadConfig().Hedera.Handler.ConsensusMessage.Addresses,
+		operatorAddress:       eth.PrivateToPublicKeyToAddress(config.LoadConfig().Hedera.Client.Operator.EthPrivateKey).String(),
```
Comment (failfmi) [inline app/process/handler/consensus-message/consensus-message.go:45]: pass config from main instead of loading it

File: app/process/handler/consensus-message/consensus-message.go [hunk 17]
```diff
@@ -79,33 +101,107 @@ func (cmh ConsensusMessageHandler) handlePayload(payload []byte) error {
...
+		log.Error("Could not submit Topic Consensus Message [%s] - [%s]", topicMessageBytes, err)
+	}
+}
+
+func (cmh ConsensusMessageHandler) enoughSignaturesCollected(txSignatures []message.TransactionMessage, transactionId string) bool {
+	requiredSigCount := len(cmh.operatorsEthAddresses)/2 + len(cmh.operatorsEthAddresses)%2
+	log.Infof("Required signatures: [%v]", requiredSigCount)
+
+	if len(txSignatures) < requiredSigCount {
+		log.Infof("Insignificant amount of Transaction Signatures for Transaction [%s] - [%d] signatures", transactionId, len(txSignatures))
```
Comment (failfmi) [inline app/process/handler/consensus-message/consensus-message.go:181]: `log.Infof("Insignificant amount of Transaction Signatures for Transaction [%s] - [%d] signaturеs out of [%d].", transactionId, len(txSignatures), requiredSigCount)`

File: app/services/signer/eth/signer.go [hunk 18]
```diff
@@ -21,3 +22,13 @@ func NewEthSigner(privateKey string) *Signer {
 func (s *Signer) Sign(msg []byte) ([]byte, error) {
 	return crypto.Sign(msg, s.privateKey)
 }
+
+func PrivateToPublicKeyToAddress(privateKey string) common.Address {
```
Comment (failfmi) [inline app/services/signer/eth/signer.go:26]: Attach func to struct Signer with name `Address() common.Address` and add signer to `ConsensusMessageHandler` struct

File: cmd/main.go [hunk 19]
```diff
@@ -30,8 +30,8 @@ func main() {
 	ethSigner := eth.NewEthSigner(configuration.Hedera.Client.Operator.EthPrivateKey)
 
 	transactionRepository := transaction.NewTransactionRepository(db)
-	statusCryptoTransferRepository := status.NewStatusRepository(db, "CRYPTO_TRANSFER")
-	statusConsensusMessageRepository := status.NewStatusRepository(db, "HCS_TOPIC")
+	statusCryptoTransferRepository := status.NewStatusRepository(db, "HCS_CRYPTO_TRANSFER")
```
Comment (failfmi) [inline cmd/main.go:33]: Use already declared variables `HCSMessageType` and `CryptoTransferMessageType`

File: app/process/watcher/util/util.go [hunk 20]
```diff
@@ -0,0 +1,30 @@
...
+	seconds, err := strconv.ParseInt(stringTimestamp[0], 10, 64)
+	if err != nil {
+		return 0, errors.New(fmt.Sprintf("Could not parse the whole part of a timestamp: [%s] - [%s]", timestamp, err))
+	}
+	nano, err := strconv.ParseInt(stringTimestamp[1], 10, 64)
+	if err != nil {
+		return 0, errors.New(fmt.Sprintf("Could not parse the decimal part of a timestamp: [%s] - [%s]", timestamp, err))
+	}
+
+	return seconds*1000000000 + nano, nil
```
Comment (Daniel-K-Ivanov) [inline app/process/watcher/util/util.go:23]: you could extract `1000000000` to variable in the util file like:
```Go
const (
    nanosecInSecond = 1000000000
)
```

File: cmd/main.go [hunk 21]
```diff
@@ -30,12 +27,12 @@ func main() {
 	db := persistence.RunDb(configuration.Hedera.Validator.Db)
 	hederaMirrorClient := hederaClients.NewHederaMirrorClient(configuration.Hedera.MirrorNode.ApiAddress)
 	hederaNodeClient := hederaClients.NewNodeClient(configuration.Hedera.Client)
-	ethClient := ethclient.NewEthereumClient(configuration.Hedera.Eth)
+	//ethClient := ethclient.NewEthereumClient(configuration.Hedera.Eth)
```
Comment (Daniel-K-Ivanov) [inline cmd/main.go:30]: I think that this should not me commented?

File: cmd/main.go [hunk 22]
```diff
@@ -54,14 +51,16 @@ func main() {
...
+		configuration,
+		ethSigner))
 
 	err = addConsensusTopicWatchers(configuration, hederaNodeClient, hederaMirrorClient, statusConsensusMessageRepository, server)
 	if err != nil {
 		log.Fatal(err)
 	}
 
-	server.AddWatcher(ethereum.NewEthereumWatcher(ethClient, configuration.Hedera.Eth))
+	//server.AddWatcher(ethereum.NewEthereumWatcher(ethClient, configuration.Hedera.Eth))
```
Comment (Daniel-K-Ivanov) [inline cmd/main.go:63]: I think that this should not me commented?

**Review by Daniel-K-Ivanov — APPROVED** [review]

File: app/persistence/status/status.go [hunk 23]
```diff
@@ -29,43 +29,43 @@ func NewStatusRepository(dbClient *gorm.DB, statusType string) *StatusRepository
 
 func typeCheck(statusType string) {
 	switch statusType {
-	case "HCS_TOPIC":
-	case "CRYPTO_TRANSFER":
+	case "HCS_TOPIC_MSG":
```
Comment (failfmi) [inline app/persistence/status/status.go:32]: You can use already declared `CryptoTransferMessageType`, `HCSMessageType`

File: app/process/handler/consensus-message/consensus-message.go [hunk 24]
```diff
@@ -85,33 +112,106 @@ func (cmh ConsensusMessageHandler) handlePayload(payload []byte) error {
 		return errors.New(fmt.Sprintf("[%s] - Address is not valid - [%s]", m.TransactionId, address.String()))
 	}
 
-	messages, err := cmh.repository.GetTransaction(m.TransactionId, ethSig)
-	if err != nil {
+	mes, err := cmh.repository.GetTransaction(m.TransactionId, ethSig, hexHash)
+	if err != nil && !errors.Is(err, gorm.ErrRecordNotFound) {
 		return errors.New(fmt.Sprintf("Failed to retrieve messages for TxId [%s], with signature [%s]. - [%s]", m.TransactionId, ethSig, err))
```
Comment (failfmi) [inline app/process/handler/consensus-message/consensus-message.go:117]: `Failed to retrieve transaction message with TxID [%s], Signature [%s]. Error [%s].`

File: app/domain/repositories/message.go [hunk 25]
```diff
@@ -3,6 +3,7 @@ package repositories
 import "github.com/limechain/hedera-eth-bridge-validator/app/persistence/message"
 
 type MessageRepository interface {
-	GetTransaction(txId, signature string) ([]message.TransactionMessage, error)
+	GetTransaction(txId, signature, hash string) (*message.TransactionMessage, error)
 	Create(message *message.TransactionMessage) error
+	GetByTransactionWith(txId string, txHash string) ([]message.TransactionMessage, error)
```
Comment (failfmi) [inline app/domain/repositories/message.go:8]: Suggestion: 
`GetTransaction(txId, signature, hash string)` <- returns a single tx message
`GetTransactions(txId, txHash string)` <- returns an array of tx messages

File: app/process/handler/consensus-message/consensus-message.go [hunk 26]
```diff
@@ -85,33 +112,106 @@ func (cmh ConsensusMessageHandler) handlePayload(payload []byte) error {
...
-		return errors.New(fmt.Sprintf("Could not add Transaction Message with Transaction Id and Signature - [%s]-[%s]", m.TransactionId, ethSig))
+		return errors.New(fmt.Sprintf("Could not add Transaction Message with Transaction Id and Signature - [%s]-[%s] - [%s]", m.TransactionId, ethSig, err))
 	}
 
-	log.Printf("Successfully verified and persisted TX with ID [%s]\n", m.TransactionId)
+	log.Infof("Successfully verified and saved signature for TX with ID [%s]", m.TransactionId)
+
+	txSignatures, err := cmh.repository.GetByTransactionWith(m.TransactionId, hexHash)
+	if err != nil {
+		return errors.New(fmt.Sprintf("Could not retrieve Transaction Signatures for Transaction [%s]", m.TransactionId))
```
Comment (failfmi) [inline app/process/handler/consensus-message/consensus-message.go:142]: `Could not retrieve transaction messages for Transaction ID [%s]. Error [%s]`

File: app/process/watcher/consensus-message/consensus-message.go [hunk 27]
```diff
@@ -45,27 +45,27 @@ func (ctw ConsensusTopicWatcher) Watch(q *queue.Queue) {
...
 	var err error
 
 	if !ctw.started {
-		if milestoneTimestamp != "" {
+		if milestoneTimestamp > 0 {
 			return milestoneTimestamp
 		}
 
 		log.Warnf("[%s] Starting Timestamp was empty, proceeding to get [timestamp] from database.\n", topicAddress)
 		milestoneTimestamp, err := ctw.statusRepository.GetLastFetchedTimestamp(topicAddress)
```
Comment (failfmi) [inline app/process/watcher/consensus-message/consensus-message.go:59]: What if err is different from `ErrRecordNotfound`

**Review by failfmi — APPROVED** [review]
